src/promql/src/functions/holt_winters.rs

```diff
@@ -0,0 +1,347 @@
+                Self::name(),
+                ts_range.value_type()
+            )),
+        )?;
+        error::ensure(
+            value_range.value_type() == DataType::Float64,
+            DataFusionError::Execution(format!(
+                "{}: expect Float64 as value array's type, found {}",
+                Self::name(),
+                value_range.value_type()
+            )),
+        )?;
+
+        // calculation
+        let mut result_array: Vec<f64> = Vec::with_capacity(ts_range.len());
+        for index in 0..ts_range.len() {
+            let timestamps = ts_range.get(index).unwrap();
+            let values = value_range.get(index).unwrap();
+            let values = values
+                .as_any()
+                .downcast_ref::<Float64Array>()
+                .unwrap()
+                .values();
+            error::ensure(
+                timestamps.len() == values.len(),
+                DataFusionError::Execution(format!(
+                    "{}: input arrays should have the same length, found {} and {}",
+                    Self::name(),
+                    timestamps.len(),
+                    values.len()
+                )),
+            )?;
+            result_array = holt_winter_impl(values, self.sf, self.tf);
+        }
+
+        let result = ColumnarValue::Array(Arc::new(Float64Array::from_iter(result_array)));
+        Ok(result)
+    }
+}
+
+fn calc_trend_value(i: usize, tf: f64, s0: f64, s1: f64, b: f64) -> f64 {
+    if i == 0 {
+        return b;
+    }
+    let x = tf * (s1 - s0);
+    let y = (1.0 - tf) * b;
+    x + y
+}
+
+/// Refer to https://github.com/prometheus/prometheus/blob/main/promql/functions.go#L299
+fn holt_winter_impl(values: &[f64], sf: f64, tf: f64) -> Vec<f64> {
+    if sf.is_nan() || tf.is_nan() || values.is_empty() {
+        return vec![];
+    }
+    if sf < 0.0 || tf < 0.0 {
+        return vec![];
+    }
+    if sf > 1.0 || tf > 1.0 {
+        return vec![];
+    }
+
+    let l = values.len();
+    if l <= 2 {
+        // Can't do the smoothing operation with less than two points.
+        return vec![];
+    }
+
+    let values = values.to_vec();
+
+    let mut s0 = 0.0;
+    let mut s1 = values[0];
+    let mut b = values[1] - values[0];
+
+    let mut result = vec![];
+    result.push(s1);
+    for (i, value) in values.iter().enumerate().skip(1) {
+        // Scale the raw value against the smoothing factor.
+        let x = sf * value;
+        // Scale the last smoothed value with the trend at this point.
+        b = calc_trend_value(i - 1, tf, s0, s1, b);
+        let y = (1.0 - sf) * (s1 + b);
+        s0 = s1;
+        s1 = x + y;
+        result.push(s1);
+    }
+    result
+}
+
+#[cfg(test)]
+mod tests {
+    use datafusion::arrow::array::TimestampMillisecondArray;
+
+    use super::*;
+    use crate::functions::test_util::simple_range_udf_runner;
+
+    #[test]
+    fn test_holt_winter_impl_validation_rules() {
```

Comment:
happy to split it into several tests

Reply:
It's ok to do so. This fn is not too long 👍 But you can put empty lines between each case.

Reply:
I decided to group assertions logically and split it into separate tests